File: medium/Daily Temperatures.js
/*
    思路:   
        這一題蠻有趣的 , 給定一個temperature , 裡面的值代表溫度
        要返回另一個array , 裡面的內容是 temperature[i] 需要等待幾天才會遇到下一個溫度更高的日子
        如果沒有這樣的日子 , 返回0 

        暴力解法就是雙迴圈去數 , 但鐵定不會過完全測資 
*/


/**
 * @param {number[]} temperatures
 * @return {number[]}
 */


// 暴力雙迴圈 
var dailyTemperatures = function(temperatures) {
    
    let solution = new Array(temperatures.length).fill(0) ; 

    for (let i = 0 ; i < temperatures.length ; i++){

        const base = temperatures[i] ; 

        for (let j = i+1 ; j < temperatures.length ; j++){

            if (temperatures[j] > base){
                solution[i] = j-i ; 
                break ; 
            }

        }
    }

    return solution ; 

};


// DP 
var dailyTemperatures = function(temperatures) {


}

File: medium/Daily Temperatures.py
""" 
    思路 :  
        這一題的概念跟next greater element基本上就是一樣 , 
        在我看來就是要多保存index而已  ,
        因為這一題要求返回的是 NGE 距離本身的index距離為何 
"""


""" 
    解法一: 單調推疊+紀錄index
    
    element in the stack : ( temperature of ith day , index  )
    
"""
from typing import List 
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        
        sol = [None for i in range(len(temperatures))] 
        stack = [] 
        
        for i in range(  len(temperatures) -1 , -1 ,-1 ): 
            
            while stack and stack[-1][0] <= temperatures[i] : 
                stack.pop() 
            
            sol[i] = ( stack[-1][1] - i ) if stack else 0 
            stack.append( (temperatures[i] , i)  )

        return sol 
            
            

        